[INTEL_MKL] Parallelizing scatter update op.

Srini511 opened the pull request:
This is a follow up for https://github.com/tensorflow/tensorflow/pull/30375 and fixes the regression failure.

INTRA_OP=1
Benchmark Time(ns) Iterations
BM_ScatterAddInt32/1 89109380 100 11.2M items/s
BM_ScatterAddInt32/10 123866710 100 80.7M items/s
BM_ScatterAddInt32/64 177304560 100 361.0M items/s
BM_ScatterAddInt32/256 270029460 100 948.0M items/s
BM_ScatterAddInt32/1024 762985280 100 1342.1M items/s

INTRA_OP=13
Running main() from test_main.cc
Benchmark Time(ns) Iterations
BM_ScatterAddInt32/1 47192100 100 21.2M items/s
BM_ScatterAddInt32/10 48498980 100 206.2M items/s
BM_ScatterAddInt32/64 52246600 100 1225.0M items/s
BM_ScatterAddInt32/256 70786030 100 3616.5M items/s
BM_ScatterAddInt32/1024 150688620 100 6795.5M items/s

Commit:
Parallelizing scatter update op.

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
+    } else {
+      num_locks = kMaxLocks;
+      entries_per_lock = (limit % kMaxLocks == 0) ? limit / kMaxLocks
+                                                  : (limit / kMaxLocks + 1);
```
This can be more concise. Please also make each comment line fill the whole line (80 characters) before beginning a new line.
```c++
const Index num_locks = std::min(limit, kMaxLocks);
const Index entries_per_lock = (limit + kMaxLocks - 1) / kMaxLocks;
```

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
+
+    std::vector<std::atomic<bool>> accessed(num_locks);
+    auto ParallelInit = [&](Index start, Index end) {
+      for (Index i = start; i < end; i++) accessed.at(i) = false;
```
Nit: 
```suggestion
      for (Index i = start; i < end; ++i) accessed.at(i) = false;
```
The same for other places.
(Just wanted the style to be consistent.)

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
+    auto ParallelInit = [&](Index start, Index end) {
+      for (Index i = start; i < end; i++) accessed.at(i) = false;
+    };
+    Index bad_index = -1;
```
`bad_index` should be atomic.

Srini511 commented:
@penpornk Gentle reminder for reviewing!

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
+        // Separating test from test and set to improve performance and reduce
+        // coherence overhead.
+        // Test
+        while (accessed.at(lock_id)) {
```
Let's make them one-liners. 
```c++
        while (accessed.at(lock_id)) {}  // Test
        while (accessed.at(lock_id).exchage(true)) {}  // Test and set.
```

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
     }
-    return -1;
+
+    std::vector<std::atomic<bool>> accessed(num_locks);
```
How much does the performance differ from `std::atomic_flag`? Could you please try [std::atomic_flag::test_and_set](http://www.cplusplus.com/reference/atomic/atomic_flag/test_and_set/) as well?

penpornk replied:
Would love to know how TF's `mutex` performs too.

Srini511 replied:
Do you mean fine grained locking with tf's mutex?

penpornk replied:
Yes, just curious.

Srini511 replied:
Mutex
BM_ScatterAddInt32/1      12911020        100    77.5M items/s
BM_ScatterAddInt32/10     14440060        100    692.5M items/s
BM_ScatterAddInt32/64     19285140        100    3318.6M items/s
BM_ScatterAddInt32/256    43886400        100    5833.2M items/s
BM_ScatterAddInt32/1024  129769780        100    7890.9M items/s

atomic_flag.test_and_set()
BM_ScatterAddInt32/1      19393250        100    51.6M items/s
BM_ScatterAddInt32/10     23358120        100    428.1M items/s
BM_ScatterAddInt32/64     32073190        100    1995.4M items/s
BM_ScatterAddInt32/256    46956580        100    5451.8M items/s
BM_ScatterAddInt32/1024  130544220        100    7844.1M items/s

Current
BM_ScatterAddInt32/1      27138230        100    36.8M items/s
BM_ScatterAddInt32/10     30097940        100    332.2M items/s
BM_ScatterAddInt32/64     35067340        100    1825.1M items/s
BM_ScatterAddInt32/256    54302060        100    4714.4M items/s
BM_ScatterAddInt32/1024  129844650        100    7886.3M items/s

Fine grained mutex seems to perform best overall. However in a model we have the difference between the 3 is not as much pronounced.

Srini511 replied:
I can change the code to either. If we are going to opt for mutex option, I would like to know how the GUARDED_BY/EXCLUSIVE_LOCKS syntax works since it is fine grained locking. Please let me know.

penpornk replied:
This is great! Thank you very much for trying!
Since `Mutex` doesn't hurt your model either, let's use `Mutex`. 

I don't think any of the [thread-safety annotations](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/default/thread_annotations.h) are applicable here. `GUARDED_BY` is used when the variable is declared.  We can't really guard `params` as it is passed into the functor. `EXCLUSIVE_LOCKS_REQUIRED` and `SHARED_LOCKS_REQUIRED` means locks must have been acquired before the function, which is not suitable for our case. 

`ACQUIRE_SHARED` and `RELEASE` can't be used together so I don't think we can tag that on the lambda function either.

Srini511 replied:
Thanks for the clarification! Do you have an insight of where the perf difference is coming from?

penpornk replied:
I'm guessing that the results are with INTRA_OP=13? Do you see much difference with INTRA_OP=1 or 2? If not, then it's probably because of a more careful implementation to avoid contention between threads.

TensorFlow's [Mutex](https://github.com/tensorflow/tensorflow/blob/9adba1c5ea0eee85dd215a9e9be51a009ce771f2/tensorflow/core/platform/default/mutex.cc#L43) uses [nsync's mutex](https://github.com/google/nsync/blob/master/internal/mu.c). Threads can sleep and wake up every once in a while to check on a lock to avoid contention.

A downside of `while(test)` following by`while(test-and-set)` is that the threads can still contend at `while(test-and-set)` (many threads getting through `while(test)` at the same time).

Srini511 replied:
Thanks for the explanation!

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
+    const DeviceBase::CpuWorkerThreads& worker_threads =
+        *(c->device()->tensorflow_cpu_worker_threads());
+    Shard(worker_threads.num_threads, worker_threads.workers, num_locks, 3500.0,
+          ParallelInit);  // Cost is arbitrary for now.
```
Please make the comment a todo instead.
```suggestion
          ParallelInit);  // TODO: Find the best cost estimation.
```

penpornk commented on tensorflow/core/kernels/scatter_op_test.cc:
```diff
@@ -175,6 +186,47 @@ TEST_F(ScatterUpdateOpTest, Error_IndexOutOfRange) {
+  MakeOp(DT_INT32_REF, DT_INT32);
+  // Feed and run
+  const int kRows = 1;
+  std::vector<int32> values;
```
For readability, we can initialize `std::vector` in the same line.
```suggestion
  std::vector<int32> values(kRows, 0);
```
Same for `indices` and `updates`.

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
+    };
+    const DeviceBase::CpuWorkerThreads& worker_threads =
+        *(c->device()->tensorflow_cpu_worker_threads());
+    Shard(worker_threads.num_threads, worker_threads.workers, num_locks, 3500.0,
```
The array size is pretty small (`kMaxLocks = 1024`). How much faster is multithreaded vs serial initialization?

Srini511 replied:
For 1024, there is not much difference. But we did observe that for some cases, larger number of locks (> 8K) gave better performance. That is why decided to leave the parallel init code in.

penpornk replied:
Got it. Thank you! Please add the explanation as a comment in the code so people who look at the code later can know.

penpornk reviewed (changes requested):
Thank you very much for re-submitting the PR!

Commit:
Addressing review changes

Srini511 commented:
Hi @penpornk, I have made the changes requested. Thanks

penpornk commented on tensorflow/core/kernels/scatter_functor.h:
```diff
@@ -188,6 +189,7 @@ struct AssignSYCL<scatter_op::UpdateOp::MAX> {
 }  // namespace scatter_op
 
 namespace functor {
+#define kMaxLocks 1024
```
Let's make this a constant variable.
```suggestion
const Index kMaxLocks 1024;
```
Then you won't have to cast `kMaxLocks` to `Index` in `std::min`.

penpornk commented on tensorflow/core/kernels/scatter_op_test.cc:
```diff
@@ -175,6 +186,46 @@ TEST_F(ScatterUpdateOpTest, Error_IndexOutOfRange) {
+  MakeOp(DT_INT32_REF, DT_INT32);
+  // Feed and run
+  const int kRows = 1;
+  std::vector<int32> values(kRows, 0);
```
The second argument to [std::vector constructor](https://en.cppreference.com/w/cpp/container/vector/vector) contains the value to initialize the whole vector with, so you can delete the for loop.

Alternatively, since array size is known at compile time, we can also use `std::array`.
```c++
std::array<int32, kRows> values;
values.fill(0);
```

penpornk commented on tensorflow/core/kernels/scatter_op_test.cc:
```diff
@@ -175,6 +186,46 @@ TEST_F(ScatterUpdateOpTest, Error_IndexOutOfRange) {
+  }
+  const int kNumUpdates = 1000000;
+  std::vector<int32> indices(kNumUpdates, 0);
+  std::vector<int32> updates(kNumUpdates, 0);
```
```suggestion
  std::vector<int32> updates(kNumUpdates, 1);
```
And you can delete the for loop below.

Or we can switch to `std::array` as well.

penpornk reviewed (changes requested):
Thank you for the changes!

Commit:
Some more review comments

Srini511 commented:
@penpornk Done!

penpornk reviewed (dismissed):
Thanks for your patience! Let's see how it goes this time. :)

Srini511 commented:
Hi @penpornk , Please let me know how the tests look. The failing test has passed this time!

penpornk commented:
Hi @Srini511, Sorry for the delay! Yes, there is a compilation error in scatter_functor.h:
```
error: variable length array used [-Werror,-Wvla]
    mutex accessed[num_locks];
```
Please either dynamically allocate the array or use `kMaxLock`.

Srini511 commented:
Hi @penpornk , how did the tests pass then?

Commit:
minor change to fix compilation error

Srini511 commented:
@penpornk Pushed the change to fix the array allocation. Thanks

penpornk reviewed (approved):
Thank you for the changes!
How the tests passed here and for you: Open-source TF uses gcc compiler. Our internal tests uses Clang compiler which is stricter than gcc.

Srini511 commented:
@penpornk Thanks! crossing fingers again!

Srini511 commented:
Hi @penpornk are the tests okay?

penpornk commented:
@Srini511 Sorry for the delay! There was a problem pulling this PR in yesterday. We are trying again now. (The internal tests will be run after the changes are pulled in.)

Srini511 commented:
Thanks @penpornk! No problem. Please let me know if there is any change needed from my end.  Has import/copybara failed?

penpornk commented:
@Srini511 Thanks for your understanding! Yes, import/copybara failure was the problem.

Srini511 commented:
Thanks @penpornk. Is it from my end?

rthadur commented:
> Thanks @penpornk. Is it from my end?

not from your end.!

penpornk commented:
No, it's from our end. In fact, it looks like `import/copybara` still has a problem. I'm going to force run the tests again just so `import/copybara` gets rerun.

Srini511 commented:
Thanks @penpornk. I looked at the new failing Ubuntu contrib tests. They seem to be spurious and not related to the PR. Can you please take a look! 
//tensorflow/contrib/layers:target_column_test is passing at my end

penpornk commented:
@Srini511 The errors are unrelated. And those tests used to pass before (without changes to the PR) so no worries! I only reran the tests in hopes to get rid of the `import/copybara` problem.

penpornk commented:
This PR slowed down a test by 2-3x. I'll have to roll it back. Sorry about this!

How to replicate:
```
$ bazel build --config=opt --test_arg=--logtostderr --test_output=streamed  //tensorflow/contrib/metrics:metric_ops_large_test
$ bazel-bin/tensorflow/contrib/metrics/metric_ops_large_test
```
Results before this PR: Ran 2 tests in 7.639s
Results after this PR: Ran 2 tests in 17.330s

Srini511 commented:
Thanks @penpornk , can you please tell me if there is anything else that we need to look at?

penpornk commented:
@Srini511 I didn't get any other regression warnings so hopefully this is the only thing. Sorry again for the inconvenience!

Srini511 commented:
No worries @penpornk! We analyzed the test and found that the slowdown is due to lock contention.
The value of limit = 201 and N =  4M in this test. This means that 4M updates are being performed on just 201 possible indices and hence there is contention. It appears like a corner case and wondering if the PR can be accepted as is for now since the implementation is functionally correct and the benchmarks give good performance. We are also working on a solution that will address similar case too. Thanks!

penpornk commented:
@Srini511 Thank you for investigating! I think we can use the original serial code when there is high potential for lock contention. Would `N / lock > 100` work for you? I'm testing it right now. It got through `metric_ops_large_test` with a pretty small slowdown.
```c++
const float const float accesses_per_index = static_cast<float>(N) / limit;
if (accesses_per_index > 100) {
  // Serial code.
} else {
  // Parallel code.
}
```

If this (or other numbers) works for you, would you like to submit the PR again or should I just make the changes directly?

Srini511 commented:
Thanks a lot for the kind gesture @penpornk. Please make the changes directly and keep the number as 1024 as the benchmark test has max(N/limit) = 102.4

penpornk commented:
@Srini511 Unfortunately, the benchmarks (ScatterAddInt*) are significantly slower with the PR than without in normal TensorFlow. I don't have much time to work on this today so this PR will probably miss r2.0. Is that okay with you? (Otherwise I can try to guard the changes with `#ifdef INTEL_MKL`.)

Srini511 commented:
No issues @penpornk , I will submit another PR.

penpornk commented:
@Srini511 Thank you! You might have to try compiling with clang too, since the problem was that the parallelized version is slower than the serial version when compiled with clang. Otherwise, I can try to look at this in several days.

I also realized that you have to put different benchmarking calls in different functions, i.e.,
```c++
static void BM_ScatterAddInt32(int iters, int embedding_size) {
  BM_ScatterHelper<int32>(iters, embedding_size, "ScatterAdd");
  BM_ScatterHelper<int32>(iters, embedding_size, "ScatterAdd", true);
}
```
Should be
```c++
static void BM_ScatterAddInt32(int iters, int embedding_size) {
  BM_ScatterHelper<int32>(iters, embedding_size, "ScatterAdd");
}
static void BM_ScatterAddInt32Large(int iters, int embedding_size) {
  BM_ScatterHelper<int32>(iters, embedding_size, "ScatterAdd", true);
}
```

Srini511 commented:
Thanks @penpornk. can you please provide me with a patch of your changes, I can apply them and debug. I have a version internally which seems to be doing okay.

penpornk commented:
@Srini511 I just wrapped `if-else` around your code to bypass that one case (metric_ops_large_test).
```c++
const int kMaxAddressesPerIndexForParallelScatter = 1024;
const float const float accesses_per_index = static_cast<float>(N) / limit;
std::atomic<Index> bad_index(-1);
if (accesses_per_index > kMaxAddressesPerIndexForParallelScatter) {
  // Serial code. (The original code.)
} else {
  // Parallel code. (Your code.)
}
```
The benchmark performance is similar to your PR so you can just debug your PR as well. 

> I have a version internally which seems to be doing okay.
Have you tried `clang`? The PR is doing fine with `gcc` for me too.

Srini511 commented:
@penpornk, To clarify ScatterAddInt* benchmarks were significantly slower when run with clang but were okay when run with gcc?
When i split the BM_ScatterAddInt32 and BM_ScatterAddInt32Large, i saw that gcc also had some regression. I added an extra condition which says, 
const bool execute_serial =
        ((N < 10000) || ((N / limit) > ser_par_ratio)) ? true : false; If you can quickly check if this is fine internally, that would be great. Thanks!
